### ISSUE TITLE
Integrate Traffic Data

### ISSUE THREAD

**Issue by jwsp1** [organicmaps/organicmaps#1160]
There is an API for traffic data for the German highways: https://github.com/bundesAPI/autobahn-api (in German).

I guess it is much effort to implement the data but it can still be a first step of an enhancement to know about heavy traffic or closed highways.

**Comment by dumblob** [comment]
This (and other countries) is an essential necessity for real world car navigation. It is also the only reason I stay with Google Maps.

**Comment by hermann-san** [comment]
OSM and has the same issue of missing traffic data. Maybe there can be some collaboration with them or some hint's on how to make some progress
https://github.com/osmandapp/OsmAnd/issues/6878

**Comment by SlouchyButton** [comment]
I'd like to add that for Czechia it is possible to retrieve the traffic information for free in XML (and DATEX II, but never heard of that) live - https://dopravniinfo.cz/cs/pages/data-from-jsdi/traffic-information. They should also contain (apart from traffic jams) all traffic incidents (crashes, closures etc.) and expected time to finish the event (finishing works on road or removing crashed cars).

I'd propose that for a sake of usability, the navigating algorithm should be changed, so it can work with traffic information and create some way in which we could implement individual traffic provider sources into the app. My knowledge with C++ and C is very limited for any in depth help with this, but I'd very gladly try to implement the Czech traffic information.

Down the road the system could be expanded with _optional_ sending of current location with speed to servers for "in-house" traffic data not depended on any 3rd party.

**Comment by biodranik** [comment]
Of course, the server should collect and pre-process traffic data, not millions of clients. Otherwise, that free public API will be DDOSed and closed.

It is a good idea to make an aggregate issue with a list of all publicly available sources/countries, so we can have a good starting point.

Are there already some aggregators for information like that?

**Comment by SlouchyButton** [comment]
@biodranik I don't think there is in this project. When researching I found one source of these traffic APIs for different countries, but it seemed outdated (was on OSM wiki if I remember correctly). I propose making a wiki page with a table where we could link all the public (and/or private) traffic APIs for future reference. I could try and look up if other countries (except DE and CZ) has something like this and make more comprehensive list.

Personally, I'd try to gather information directly from the official sources of individual countries. I expect most of them should be free, since it is official source. Going the route of paid plug-in, or subscription service doesn't seem a viable option, because there are many alternative free navigation apps. Having this pretty crucial feature behind paywall would inevitably hurt the app.

You mentioned a server collecting this data and processing it. If needed, I can provide server for backend as I have business in web dev, and I own a physical server located in the Czech Republic, Brno. (Have no idea what are current options for backend servers for this app)

Regarding the OsmAnd and OrganicMaps collaboration (as proposed by @hermann-san ), there is one possibly mutually beneficial way of collaborating. If both apps implemented the _optional_ location and speed tracking for their users, both apps would have much bigger user base together. This would result in much more accurate information about traffic, than would be possible to achieve only relying on the user base of one of the apps. I would understand any privacy concerns regarding this feature, but this could be implemented in a privacy respecting way and on an optional basis. Together with open source nature of app, it could be fully transparent.

**Comment by dumblob** [comment]
Hey everybody!

I just stumbled upon an open source project which seems to actively gather real-time traffic data in a multinational manner.

https://traffxml.gitlab.io/ (https://gitlab.com/traffxml )

Thery also have an Android app [RoadEagle](https://gitlab.com/traffxml/roadeagle ).

Could you get in contact with them ( info@traffxml.org ).

Btw. [Navit](https://github.com/navit-gps/navit ) seems to use their traffic data. @metalstrolch @bkoppelmann @jkoan @OLFDB @mvglasow and others could maybe know something more.

**Comment by mvglasow** [comment]
Yes, TraFF is my project (so far I’ve been alone on this, but contributions will certainly be welcome).

Just to give you an idea of what TraFF is: at its core, there is an open data format for traffic information, designed for fully machine-parseable event information, independent of the underlying map.

On the software side, there is a library to parse and generate feeds in the TraFF XML format, as well as import libraries for a bunch of formats, including the Datex II format, which is widely used in the EU.

As for the software ecosystem, so far there are three workable products:
* Navit, a navigation system which processes TraFF data (this feature is currently available on Android only)
* RoadEagle, an Android app which reads traffic information from a number of sources and makes them available on screen as well as in TraFF format (currently with support for Poland, Lithuania and Latvia)
* Qz, another Android app which can receive TMC data via an external hardware device and makes it available on screen as well as in TraFF format (works in a number of places in western an d central Europe)

A TraFF server (which would aggregate data from various sources and make them available in TraFF format via an API) is in the works. Future clients could then directly connect to a TraFF server, rather than tap into sources locally. This would also allow for integration of a bunch of additional sources, which are not feasible to process with an Android app (e.g. the Datex-II library relies on components not available on Android, and the Czech service mentioned above requires running a server at a static IP address).

The source code for all of this can be found at https://gitlab.com/traffxml (the only exception being Qz, which lives at https://gitlab.com/mvglasow/qz.)

Main obstacle to implementing all of this is my own limited free time – as I said, if anyone would like to participate in the further development of TraFF, I’d be more than happy.

As for the German API, I have come across that before. While its authors look official (“Bundesstelle für Open Data”), they are actually an activist organization. I am not sure where exactly they pull their data from and under what terms. It might be a legal gray area in the sense that they do not have a formal license to use the data, but since all it was ever intended for is to power a free-of-charge app, nobody has objected to this usage so far.

However, Germany does offer some sources of traffic data which are officially free to use. Visit https://mcloud.de and search for “Datex”, which will give you a bunch of sources in Datex-II format. (The TraFF Datex-II converter will convert that into TraFF, though you might need some external data to decode the locations.)

As for TraFF, there is a list of potential sources at https://gitlab.com/traffxml/roadeagle/-/wikis/Data-Sources.

**Comment by dumblob** [comment]
Potentially related: https://github.com/organicmaps/organicmaps/discussions/4102 (Overture Maps Foundation - seems governed by Linux Foundation).

**Comment by gerion0** [comment]
> As for the German API, I have come across that before. While its authors look official (“Bundesstelle für Open Data”), they are actually an activist organization. I am not sure where exactly they pull their data from and under what terms. It might be a legal gray area in the sense that they do not have a formal license to use the data, but since all it was ever intended for is to power a free-of-charge app, nobody has objected to this usage so far.

The "Bundesstelle für Open Data" only documents the API for data that are provided officially by the government.
In this case let me translate the README of the link [Autobahn App API](https://github.com/bundesAPI/autobahn-api):

> The BMVI (official department) built a completely useless app. An autobahn information app
> See: https://autobahn.api.bund.dev/

> Really completely useless?
> No, since there is a public API for up-to-date data with regard to construction side information, webcams, parking places, ... - surely useful for someone.

So to say, while the API documentation is inofficial, the information itself comes directly from autobahn.de which is a official website of the Autobahn GmbH which belongs to the government.

**Comment by xephos1one** [comment]
This might be useful:
https://github.com/graphhopper/open-traffic-collection

**Comment by dikkechill** [comment]
> 

The above mentioned TraFF project (https://gitlab.com/traffxml) provides quite a number of building blocks already:
1. Open data format specification (TraFF XML) to normalize/standardize various data feeds to
2. Mappings from common public data feed formats (e.g. Datex II) to TraFF XML format
3. A server implementation for collection and pre-processing public data (TraFF server) with proof of concepts for Belgium, Germany, Finland, Latvia, Lithuania, Luxembourg, the Netherlands and Poland

For the longer term this still requires:
* TraFF XML implementation in OrganicMaps
* Server(s) running for OrganicMaps (and maintenance of the infrastructure)
* Support on TraFF project (e.g. adding and updating data sources, ) -  seems currently a one person who has done a lot of work

Maybe it's possible to build a PoC with the current traff-server (or a sample traffxml file) and a traffic layer (view only) in OrganicMaps?

**Comment by biodranik** [comment]
Thanks, that's an interesting project.

1. What are pros/cons of TraFF vs other implementations (if there are any)?
2. Why only one person is working on it? Who else supports it?
3. What are server requirements?
4. How server handles millions of clients and a large number of requests all over the world in busy peak hours? How does it scale? Are there any load tests/performance results?
5. How many users are already using it? Can we share a server with many other project to get a better-quality data?
6. Who will own this data?

**Comment by mvglasow** [comment]
Happy to hear you’re interested in my project. Regarding your questions:

### Pros/cons

Pros:
* Fully machine parseable and language neutral events: Events are expressed through predefined codes rather than freeform text – unlike Bing and Waze which rely on freeform text for most of the details. This makes them easy to parse and easy to present in any language – it is just a matter pf providing a lookup table of predefined strings to display a TraFF message in any language.
* High level of information detail: A message (i.e. a report about a situation affecting one particular stretch of road) can be expressed by any combination of some 200 event codes, some 200 qualifiers as well as quantifiers for numeric values. Events would tell you that there is a roadworks area, with a reduced number of lanes and a temporary speed limit; quantifiers would tell you how many lanes are closed and what the speed limit is. Again, Bing and Waze do so only in very rudimentary form, and much of the important information is only available as freeform text.
* Easy-to-parse location and direction information: A location is primarily expressed through 2–3 points in WGS84 coordinates, as well as directionality information (one or both directions). Other attributes, such as road class, road number/name and geographical area, can be added but are only meant as auxiliary information. WGS84 coordinates are a must, as they are presumably well understood by any GPS-based navigation system. In comparison to other locations:
  * TMC requires recipients to obtain a location table in order to decode locations. While this saves bandwidth, it makes the information useless without the location table and limits the locations which can be encoded.
  * Some sources (e.g. autobahn.de) do not designate both message endpoints: either one endpoint is missing, or it is not clear which direction the message refers to.
  * Other sources (e.g. Datex-II) offer encoders a lot of flexibility to use their preferred location scheme (options including TMC-style location tables or coordinate pairs in the encoder’s preferred coordinate system), but that puts a massive burden on the processing end.
* Built on top of a well-defined data format – XML is one of a handful of standard formats. The schema is intentionally kept simple – in contrast to, say, Datex-II.
* Conversion libraries are available for a number of other formats (Datex-II as well as some custom ones; TMC has also been converted, albeit the code is not in library form at the moment).

Cons: Adopting TraFF is going to be pioneering work. The project is still in an early stage, not last because it is currently one guy (myself) doing this in his spare time. However, any help will be very much appreciated and I will be happy to support adoption into a widely used project.

### Community

As mentioned above, so far no one has come forward with major contributions. The ecosystem so far consists of:
* Navit, a navigation app which has support for TraFF data (on Android, it can receive data from Qz and RoadEagle)
* Qz,. an Android app which connects to a Bluetooth TMC receiver, retrieves messages, converts them to TraFF and feeds them to apps
* RoadEagle, an Android app which retrieves traffic information for (currently) Poland, Lithuania and Latvia (which use custom formats) and feeds them to apps

It appears to me that most FOSS-minded people don’t own or drive cars, and car drivers tend to go for what’s available – which is mostly proprietary at this moment.

This is why I have been reaching out to navigation apps, as I expect they have the user base with the biggest interest in that kind of data, while also being in a position to contribute.

### Server

Word of caution: the server is still in early development stage (the conversion libraries and Android apps are more mature), and no load tests have been performed yet. Requirements are:
* Java servlet engine (tested with Tomcat)
* Database backend (development is currently on HSQLDB, with plans to move to PostgreSQL)
* Linux OS preferred, others are untested

This also means there is no live server yet. If OrganicMaps can help out here, that would be great. Ultimately, sharing a server between multiple projects is encouraged as expenses would be shared while the benefits would be multiplied.

On the other hand, the feed format and conversion libraries do not depend on the server. Also, the TraFF specification includes a local delivery mechanism for Android – one app connects to a non-TraFF backend, converts messages to TraFF and announces them via a broadcast, which any app with support for it can pick up.

### Data ownership

The data sources I am using so far are under various open data licenses, some with and some without a share-alike clause. I would like the data to be available in a spirit similar to OSM itself. That primarily means the information should be free to use and redistribute. Share-alike is desirable as long as it doesn’t limit our access to data sources.

**Comment by biodranik** [comment]
@mvglasow Thanks for the detailed answer! More questions:
1. How suitable is this approach to collecting data from drivers all over the world, and providing both: a) live information that helps to build a faster route right now, and b) statistical data from the past for better route planning? How hard would it be to implement it? Are there any existing projects/source codes for these tasks?
2. Why Java for the server?

**Comment by mvglasow** [comment]
Crowdsourcing data was the original idea and TraFF was developed with that in mind. Tapping into other sources was more of a quick win so one could start out with a usable set of traffic information at least in some regions (and even with crowdsourced data, I envision the two kinds of sources might very well complement each other).

As of TraFF 0.8, the mechanism for users to report traffic data back to a central repository is not implemented, simply because I haven’t gotten around to doing it, but this is definitely planned for future versions.

One point to address in this case is authentication: the current “downstream-only” version of TraFF does not require any permanent user ID. When users wish to receive updates, they subscribe to a certain set of areas and road categories and receive a randomly generated subscription ID, which is valid only for this particular subscription (typically for one trip, but ultimately controlled by the subscriber side). The first feed contains all currently active messages matching the selected criteria, after which only “delta” feeds are sent to save bandwidth. As subscription areas identify where the user currently is and/or intends to travel, associating them with a natural person should be prevented. By using a random ID which changes with every new subscription, data exposed to the server is essentially just the IP address, timestamp and subscription area (all of which are pretty much unavoidable). Also, the ID cannot be used to compile a ”history” of a certain person’s travels.

With user-supplied messages, this may change as we may need a way to keep out the bad guys – anything from sloppiness to outright malicious insertion of bogus data. At the bare minimum, we can block a known bad user (or silently ignore their messages) if we’re alerted to some form of abuse. We might also consider some form of reputation: a message which is confirmed by other sources increases that user’s reputation; if a message is at odds with what other sources report, it decreases it. That way, when correlating messages from multiple users, we can give more weight to messages from trustworthy users.

At the same time, we still want to minimize privacy impact. I suggest that even for authenticated users, subscriptions should not be associated with a user ID (only user-submitted messages should), and any association between a user-supplied message and the user ID should be discarded when the situation to which the message refers has been resolved.

Another point will be correlating data from multiple sources but referring to the same stretch of road. Ideally, the server should detect when multiple messages affect the same stretch of road, make sense of the information and process this into one single message. For example:
* An official source reports roadworks on a certain stretch of road and a driver in that same section reports a tailback, with the speed at which traffic is flowing and the coordinates where the tailback ends.
* One driver reports a traffic jam they got into, with an accurate speed at which traffic is flowing and the end of the tailback when they encountered it. Another driver on the opposite carriageway (where traffic is flowing freely) sees the obstruction and reports where the “tip” of the queue is. Together, these two messages will give us the length of the tailback, along with a speed estimate.
* Multiple drivers in the same queue report their position and speed, and the server concludes they refer to the same event and bundles them into the same message.

None of this is implemented yet – at the moment, the server will simply collect and forward messages. Implementing the above will be another milestone and will require some kind of algorithm capable of answering the following questions:
* How do we determine two messages from different sources refer to the same event? For example:
  * Locations overlap
  * Both messages refer to a stretch of road enclosed by two junctions (making it impossible to encounter the location of one message but not the other), so they probably refer to a situation which extends across both locations
* How do we deal with two related messages containing different information?
  * If the information is complementary (e.g. one message reports roadworks, the other a queue), combine it (the resulting message will report a queue due to roadworks)
  * If there are contradictions, figure out how to resolve them:
    * Detect bogus data (a driver stopping on the shoulder due to a vehicle defect is misidentified as being in a queue)
    * The situation may have changed over time (newer message wins)
    * Some parameters of the situation may vary within the location (traffic on a long affected stretch may flow at different speeds in different places, extrapolate and/or calculate an average value)

If we keep past data, this algorithm could be implemented into generate a forecast (e.g. traffic jams tend to happen Mondays to Fridays, during the morning and evening peak hours). Though I would note that this can get quite complex: we’d have to sort out anomalies (e.g. traffic accidents or roadworks slowing down traffic more than usual, public holidays, school holidays), as well as consider upcoming anomalies (e.g. scheduled roadworks or holiday schedules). If we get an actual report from a location for which we have a forecast, we need to figure out how to correlate them.

The TraFF transport format currently does not provide an explicit mechanism for forecasts, but that could be added. The important thing would be to clearly mark forecasts as such, so that legacy receivers will not mistake them for actual reports. Also, bandwidth will need to be considered, as only a small subset of all road stretches will have actual messages at any given time, thus a comprehensive set of forecasts for each and every stretch of road will likely be at least an order of magnitude larger. (Location references in TraFF are currently for a single stretch of road; there is no such thing as, say, “all national roads in Poland”.)

To my knowledge, the only existing source code for collecting events are the TraFF server as well as the Android apps, and none of these currently has capabilities to merge events or generate forecasts. In Navit I was kind of forced to do this (we ultimately have to match an event to the map and stick attributes to it). Here, events translate to “traffic distortions”, which can result in a lower speed (0 to indicate closure) and/or an extra delay for the affected segments. The (admittedly crude) solution is to assume the lowest speed and longest delay of any traffic distortion referring to a given road segment.

As for Java on the server, that was a quick win related to how the ecosystem evolved. The Android apps (Qz and RoadEagle) came first, and with RoadEagle came the conversion libraries – which are in Java. Implementing the server in Java allowed me to use the same conversion libraries for both the Android app and the server.

None of this is set in stone – if someone comes up with a better server implementation, I’m fine with that as long as the new implementation provides the same API. Server architecture could be reevaluated as well – it might make sense to have a core server which just manages messages and subscriptions, with third-party message retrieval/conversion being done by services external to the server (and written in a different programming language than the server) and fed into the server through a yet-to-be-defined API. (This API can be server implementation-specific, or, once we have an API for reporting user data, that could also be used for server-to-server communication.)

**Comment by mvglasow** [comment]
Any news on this? If you were to tell me TraFF and any ecosystem around it is still very much in its infancy, I would definitely agree. However, it’s sort of a chicken-and-egg problem: without a user base, few developers will come forward with contributions, there won’t be a workable solution for a long time, and until then, there is no user base and the circle closes. And the whole thing is kinda large for a single person to tackle. In order to get something workable off the ground, we’d need a few people joining forces and coming up with a minimum workable solution, so we can start growing a user base and attract more contributors. I’ll be happy to contribute, to the extent time permits. However, I will definitely need some help on extending Organic Maps to process traffic information.

**Comment by mvglasow** [comment]
Great to hear that some functionality seems to exist already. How familiar are you with the code base? (Or who else can help with that part of the code?)

What I’d suggest for the PoC:
* Use one source (number of countries/areas then doesn’t matter).
* For traffic events, start with slow traffic/events and closures. This is what I also implemented for Navit and is probably the most straightforward thing to implement.
* For the source and interface, RoadEagle and subscription via Android intents and content providers would be a rather quick win. RoadEagle already has everything in place for that. And unlike a local XML file, we would get some real data right away. Only limitation: RoadEagle is available on Android, and a transport interface for iOS (or desktop Linux) has yet to be defined.
* Visualization as a first step (but I’d like to move on to routing in the next step – depends on what is already in place.)

**Comment by mvglasow** [comment]
Just looked around the code a bit and found two things:

* There seems to be OpenLR support (in the openlr folder), though I haven’t yet figured out where it is used. My first suspicion is that it might have to do with mapping the locations of traffic reports to the map (as the provider may have encoded the location using different map data, with slightly differing geometries). But that is just a guess from someone who is taking a first look at the code.
* The routing algorithm seems to be A* (as suggested by file names in routing/base). There is an extension of A* called [LPA*](https://en.wikipedia.org/wiki/Lifelong_Planning_A*) (Lifelong Planning A*), which essentially adds the ability to change the cost of select edges and do a partial recalculation of the graph. We would need that to recalculate the route when the traffic situation changes – unless this is already implemented.

**Comment by biodranik** [comment]
A great start would be to:
1. Learn the existing code base and see how it works by providing some test/stub data.
2. Document it (what's there, what's missing, which data format is used from a server).

Then we can start with a server part and a client part in parallel.

**Comment by xephos1one** [comment]
Traffic data from https://github.com/bundesAPI/autobahn-api looks like to be by TomTom Intermediate Traffic.
So the Data may only be allowed to be used inside the Autobahn App?

See https://mobilithek.info/offers/573189053963235328
And the linked details to the traffic data ("Verkehrsfluss") in the old MDM portal that got discontinued this year:
https://web.archive.org/web/20231218115539/https://service.mdm-portal.de/mdm-portal-application/publDetail.do?publicationId=2841000
I saved that page end of last year as I didn't find it in the new Mobilothek and that still links to it.
What we can see is that it's DATEX II Version 2.x using OpenLR as the Geo reference method.

Sadly I wasn't able to register to see contract information and valid from/to...

**Comment by zyphlar** [comment]
Some random thoughts:

I think it makes sense to contribute to something that's OSM-compatible in terms of licensing, as well as OM-compatible in terms of privacy and offline possibility

For example Google would be silly to route based on live data if there were only a few live data points. At some level it makes sense to modify routing based on average speed, not current speed. So we could be strategic and have a historical average data layer that either modifies estimated road speeds in the map files, or that is downloaded and parsed by the router separately

And then for live traffic incidents, rather than absolute speeds or slowdowns we could make routing logic simple by thinking of traffic as a % modifier to the road's mapped maxspeed or the measured average speed. Seems like routing based on single numbers is easiest

**Comment by auge8472** [comment]
@zyphlar The big task here is to provide the source of those data (both, historical average speed as well as current speed). Either OM (and maybe also other similar apps) feed a collector to build a data base or it is possible to soucre some of the databases that was mentioned above. Finally, data on traffic at different times on different days in different months is needed to improve the prediction of journey times. 

At the moment, it looks to me as if the app calculates the journey time using only the permitted speeds and the distance travelled. My journey to and from work in a big city is 15km and the app estimates it to take about 24 minutes. Even at 2 a.m. with only green traffic lights, this is hardly possible, let alone at any other time of day. My real journey time during the day is between 45 and 65 minutes.

**Comment by biodranik** [comment]
The current inaccurate ETA is caused by missing penalties for turns and traffic lights. #1845

**Comment by auge8472** [comment]
> The current inaccurate ETA is caused by missing penalties for turns and traffic lights. #1845

I have already surmised this as part of the problem in a comment on another issue. 🙂

**Comment by Korb** [comment]
> Google would be silly to route based on live data if there were only a few live data points.

**Google** has the largest array of data on both operational, current conditions, and hourly statistics by day of the week for most road sections. Even in countries where the company is not represented. Thus, for long routes, Google can plan a route not only based on the current traffic situation, but also based on the expected traffic load. The latter is very important for routes longer than an hour.

The very first link in Google leads to [the Traffic Layer | Maps JavaScript API](https://developers.google.com/maps/documentation/javascript/examples/layer-traffic). Could this be used for route planning?

Will this make the app fully "organic"? Will this improvement not cause the app to be flagged in F-Droid as containing things "you might not like"?

Also, I would like to point out that in countries where the majority of the population speaks Russian, **Yandex**' maps are very common — [JavaScript API Reference / traffic.provider.Actual](https://yandex.com/dev/jsapi-v2-1/doc/en/v2-1/ref/reference/traffic.provider.Actual). Honestly, I don’t even know what’s worse for paranoids: a global corporation or a local company run by bandits, whose user data is freely available on the darknet?

I would like to point out separately that drivers need traffic information not only for automatic route planning, but also for **coloring roads** in car-navigation mode.

**Comment by 1ec5** [comment]
> For example Google would be silly to route based on live data if there were only a few live data points.

Inevitably there would need to be a low-pass filter of some sort, even if we’re talking about historical data and only for route line coloring. Otherwise, isolated traffic probes can throw off route coloring or routing in very niche contexts, such as a garbage truck trudging down an alley, a security vehicle checking up on an office parking lot at night, or a dirty car being guided through an automatic car wash. The user impact would be low but wouldn’t inspire confidence in any traffic congestion coverage.

> Thus, for long routes, Google can plan a route not only based on the current traffic situation, but also based on the expected traffic load. The latter is very important for routes longer than an hour.

To set expectations, most route planners work with edge weights evaluated across the entire graph consistently as of a single instant in time. The ideal would be to instead cost each edge based on when the user is expected to arrive at that edge, but the added complexity is unnecessary for shorter routes that tend to be more common. This goes for any kind of time-dependent weighting – not only traffic congestion but also conditional restrictions and speed limits (#3731).

**Comment by geoglorus** [comment]
What about the traffic in France ? Lot’s of API are in « open data ».
https://trafic-routier.data.cerema.fr/acces-aux-donnees-de-trafic-routier-open-data-r25.html

**Comment by PureTryOut** [comment]
Has there been any progress on TraFF support? Reading this thread it seemed like the most promising solution (good job with it @mvglasow!) but there hasn't been a word about it since September 2023 and since then people talking in here have been ignoring that solution and talking about other options that I don't think are feasible (legally question use of Google data for example).
I know several people that wouldn't switch navigation apps if they didn't account for live traffic data so it seems to need to be quite a high priority issue.

**Comment by mvglasow** [comment]
I have been working on TraFF support for some time. As it turns out, the old MWM traffic code is still there and functional (just needs to be enabled in the code base), except there is no server to supply data (if there has ever been one; that would have been before my time). The data format it expects is strictly bound to a particular MWM file version, therefore parts of the code need heavy rewriting and a bunch of new features in order to support TraFF. More detail on the how and what can be found at https://codeberg.org/mvglasow/organicmaps/wiki/Traffic.
The one thing we don’t have is a central data source. The TraFF spec has the nuts and bolts for a server with a HTTP API, but the one I wrote back then is still in its infancy and the design might not be the best. There are two companion apps, both of which are Android-only, which use Binder transactions to supply traffic data. I am planning to add support for these to the Android port – haven’t gotten to that part yet. On the long run, we probably want a central TraFF server, which can be shared between multiple navigation apps. Doing so would increase the user base and thus also the base of potential contributors. In the first phase, the server would just get official traffic information from road operators who make this information available. In a second phase we could also process user contributions, but that would require additions to the TraFF transport protocol.

**Comment by zyphlar** [comment]
MWM file spec for traffic sounds like it was baked into the map generation process, and thus supporting offline statistical data more than real-time. I wonder if hooking that again could help with the offline traffic goal?

**Comment by mvglasow** [comment]
The file format is not documented, we’d either have to figure it out from code or use existing code to generate it (there are functions to serialize and deserialize traffic data from files). The server component might have been proprietary and its source code never released, it doesn’t seem to be in the repository.
Functionally, traffic data is a set of key-value pairs for each MWM: keys are segment IDs with direction, values are speed groups. A speed group represents the ratio of actual/expected speed to the posted speed limit, divided into six bins, from `G0` (stationary traffic) to `G5` (traffic flowing normally). There is also a `TempBlock` speed group, indicating a closed road.
At the file level, keys and values are in separate files. Keys are fetched only once (as they never change), values are fetched every minute.
One weakness of the model is that the data is only usable in conjunction with a particular version of the map data. That means one set of data needs to be kept for every supported map version, and other versions of the map cannot use traffic data.
If we want to work with forecast data (based on statistics, downloaded in advance, say, with the MWM file, and then used offline), we need some notion of a schedule. For example, some roads have a speed group of `G3` at peak times, `G4` at the start and end of peak times and `G5` otherwise, and peak times may differ between roads. The current data model assumes that traffic data is just a snapshot of the current situation, good for the duration of the current route, and only until an update is received. One way or the other, some new functionality will be needed to get any sort of working traffic functionality (the other alternative would be to build something functionally equivalent to the MWM traffic server, but that is a design I would not recommend).

**Comment by biodranik** [comment]
@mvglasow thanks for the detailed digging of the mapsme traffic implementation! Contrary to "patch the existing implementation", what would be the best/most flexible approach to implementing traffic info _as if it was implemented from scratch_, taking into account:
- On-device routing (with potentially different/faster algo like [OSMand's](https://osmand.net/blog/fast-routing/))
- Alternate routes
- User-reporting of `TempBlock`
- Store different types of `TempBlock` for better UX
- Rewriting current client implementation/format for more efficiency
- Writing (if necessary) a scalable server

?

**Comment by mvglasow** [comment]
> On-device routing (with potentially different/faster algo like [OSMand's](https://osmand.net/blog/fast-routing/))

OM currently uses A* for routing; the evolutionary approach would be [LPA*](https://en.wikipedia.org/wiki/Lifelong_Planning_A*) (Lifelong Planning A*). That said, I am not an expert on routing, much less on the OM implementation. There seems to be some functionality for adjusting routes (“adjust” being the term used in function names) in OM, but I doubt this is LPA*, seems more like some custom extension.

> Alternate routes

(caveat: I am no expert on routing) With LPA* instead of A*, there is still one optimal route, the only exception being that it changes as segment costs change due to traffic. This is how I implemented this in Navit.

If multiple routes are desired, one could calculate one route with and one without traffic information, then diff the two. One could then inform the user that a different route was chosen because of traffic problems (followed by a list of those traffic problems which were avoided). Though that roughly duplicates the cost of routing and adds the cost of analyzing routes and traffic problems avoided.

> User-reporting of `TempBlock`

I once had the idea of “say no to the next maneuver”. One could tap the maneuver, which would open a dialog similar to the following:

```
Avoid this maneuver? Why?

[ Road is closed/blocked ]   [ Traffic jam ]   [ Do not want ]   [ Cancel ]
```

The answer applies to the path starting at the next maneuver and up to the following junction (possibly with some adjustments to the end of the affected section). The first two would submit a report to the server. The third one would avoid the route but not report anything (user has chosen a different route out of preference). The fourth one would close the dialog without doing anything else.

The only limitation is that it would only address points after a maneuver. Following the same road usually does not generate maneuvers. One could introduce a “get me out of here” button that presents a similar dialog, but that would need more care in defining the start and end. For example, we wouldn’t want to route people out of a traffic jam on the motorway into a service area and back into the traffic jam, but guide them to the next exit.

One could also display a pop-up if users leave the route, similar to:

```
You have left the route, is this on purpose?

[ Yes, road is closed/blocked ]   [ Yes, traffic jam ]   [ Road layout changed ]   [ No, took the wrong turn ]
```

The first two behave similar to the previous example, generating a report where the user left their designated route. The third one could be used to improve the map (start tracing until the user returns to the route, insert an OSM note, ...) whereas the last one would behave as OM does now.

> Store different types of `TempBlock` for better UX

TraFF has different event types. One could link affected segments back to the TraFF report so the user would see the event. In TraFF, a message can have multiple events, and there would usually be “road closed” or “road blocked” along with the reason (landslide, flooding, public event, authority actions etc.)

> Rewriting current client implementation/format for more efficiency

Other than what I said above, pretty close to https://codeberg.org/mvglasow/organicmaps/src/branch/traffic. Current client implementation is very much based on one prebuilt set of segments per MWM, whereas TraFF is built on a user-defined subscription, which has one or more rectangles, and optionally a minimum road class per rectangle. State information in `TrafficManager` should be based on sources (there me be multiple sources, with overlapping coverage areas) rather than per MWM.

TraFF locations are map-neutral, so they need to be decoded into map segments. There is an OpenLR decoder in the code base, but it is not perfect for the task (among others, TraFF works with endpoints and optional attributes, whereas OpenLR has a reference point at each junction) and does not perform well for locations above ~10 km in length. I got far better results from a slightly modified routing engine, also in the repo linked above. A more modular router, designed to work both for route guidance and for TraFF location decoding, would be ideal.

Traffic info, along with decoded segments, should be cached to a file and reused in later sessions. Invalidating would mean discarding decoded locations for messages and regenerating them, ideally in a smart way (e.g. if the enclosing rectangle of the location overlaps with that of an updated or newly downloaded map).

> Writing (if necessary) a scalable server

See our conversation from two years ago in this ticket. My server implementation hasn’t been under active development since, and Tomcat/Java is not necessarily the best solution – I am by no means a web developer, and Java is what I know best. I would suggest roughly the architecture as described in https://gitlab.com/traffxml/traff-server/-/blob/master/README.md?ref_type=heads, and implementation similar to the code in that repo when it comes to database structure and/or how to keep track of which client has received which message.

So far I have only looked into the first stage of getting traffic information from road operators. User contributions are one step further down the road and would raise a bunch of further points to address, such as balancing privacy concerns versus protecting the information from vandalism.

Sharing the server between different projects would definitely help – it would mean a bigger user base, both in terms of message contribution and code contribution.

**Comment by 1ec5** [comment]
> One could also display a pop-up if users leave the route

Intuitively, I’d expect the user to be a little too preoccupied by this point to consistently respond. You might also get a lot of noisy data because people leave routes for all sorts of reasons. However, in aggregate, a concentration of route deviations could point to a data or guidance problem, analogous to how GPS traces in aggregate can suggest a misaligned road or missing one-way street.

**Comment by mvglasow** [comment]
Sure, that was just an idea I had while writing my answer. It also depends on the user experience – a modal dialog box in the middle of the screen is probably not good, but it could be implemented as an unobtrusive element at the top or bottom of the screen (or in a corner) which goes away if not tapped for a certain time, or if the user returns to the route.

**Comment by xephos1one** [comment]
Maybe live traffic data could be obtained from them until a production ready open source solution exists?
https://www.nunav.net/lp/sdk
Looks to be proprietary though but maybe easier than asking TomTom or so...

**Comment by razor74** [comment]
TomTom traffic is the best and i think they have the cheapest  subscription fee for this service. MapFactor Navigator already implemented traffic data with a 12 euros/year subscription for who want to pay for traffic. TomTom could provide better_than+waze traffic in all countries from Europe and even in Greece and Turkey.

**Comment by biodranik** [comment]
Related discussion: https://github.com/orgs/organicmaps/discussions/568

**Comment by The-Nickname** [comment]
Hi, the [Polar.sh](https://polar.sh/organicmaps) link isnt working anymore.
